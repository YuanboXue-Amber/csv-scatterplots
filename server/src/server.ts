import express from 'express';
import multer from 'multer';
import cors from 'cors';

const server = express();
server.use(cors());

// store file in public folder with name date+originalName
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, 'public');
  },
  filename: (req, file, cb) => {
    cb(null, Date.now() + '-' + file.originalname);
  },
});
const upload = multer({ storage }).single('file');

server.get('/', (req: any, res: any) => {
  return res.send('Hello Server');
});

server.post('/upload', (req: any, res: any) => {
  upload(req, res, (err) => {
    if (err) {
      console.error(`Error uploading file ${err}`);
      return res.status(500).json(err);
    }

    if (req.file) {
      console.log(`File uploaded to ${req.file.filename}`);
      return res.status(200).send(req.file);
    }

    return res.status(500);
  });
});

server.listen(8000, () => {
  console.log('Backend server running on port 8000');
});
