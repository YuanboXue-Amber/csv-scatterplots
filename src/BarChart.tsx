import React, { Component } from 'react';
import * as d3 from 'd3';
// tslint:disable-next-line: no-var-requires
const dsv = require('d3-dsv');

export class BarChart extends Component<{}, {}> {
  componentDidMount() {
    this.draw();
  }

  async fetchFile() {
    const response = await fetch(
      'https://raw.githubusercontent.com/thegazettedata/d3charts/master/projects/cr-bike-trails/data/letter-frequency.csv',

    );
    // Check if the request is 200
    if (response.ok) {
      return response.text();
    }
    return Promise.reject(response);
  }

  // Draw BarChart on div with id "BarChart", replace if it exists
  async draw() {
    // const data = [4, 8, 15, 16, 23, 42];
    // const csvString =
    // `name,value\nLocke,4\nReyes,8\nFord,15\nJarrah,16\nShephard,23\nKwon,42`;

    const csvString = await this.fetchFile();

    const rawData = d3.csvParse(csvString, dsv.autoType);
    const data = rawData.map<{ name: string; value: number }>(a => {
      return { name: a[rawData.columns[0]], value: a[rawData.columns[1]] };
    });

    const width = 800;
    const height = 600;
    const svg = d3
      .select('#BarChart')
      .selectAll('svg')
      .data([1])
      .join('svg');
    svg
      .attr('height', height)
      .attr('width', width);

    const margin = {top: 20, left: 30, bottom: 20, right: 20};

    const xscale = d3
      .scaleBand()
      .domain(data.map(a => a.name))
      .range([margin.left, width - margin.right])
      .padding(0.1);

    const yscale = d3
      .scaleLinear()
      .domain([0, (d3.max(data.map(a => a.value)) ?? 0 ) * 1.2])
      .range([height - margin.bottom, margin.top]);

    const xAxis = (g: any) => { g
        .append('g')
      .attr('transform', `translate(0, ${height - margin.bottom})`)
      .call(d3.axisBottom(xscale));
    };

    const yAxis = (g: any) => { g
        .append('g')
      .attr('transform', `translate(${margin.left}, 0)`)
      .call(d3.axisLeft(yscale).tickFormat(d => `${(d.valueOf() * 100).toFixed(0)}%`));
    };

    const ytitle = (g: any) => { g
        .append('text')
      .text('Frequency ^')
      .attr('font-size', xscale.bandwidth() / 2)
      .attr('transform', `translate(0, ${xscale.bandwidth() / 2})`);
    };

    svg.call(xAxis);
    svg.call(yAxis);
    svg.call(ytitle);

    const gbars = svg
      .append('g')
    .selectAll('g')
    .data(data)
    .join('g');

    gbars
        .append('rect')
      .style('fill', 'steelblue')
      .attr('height', d => yscale(0) - yscale(d.value))
      .attr('width', xscale.bandwidth())
      .attr('x', d => xscale(d.name) ?? 0)
      .attr('y', d => yscale(d.value));

    gbars
      .append('text')
      .attr('font-size', xscale.bandwidth() / 2)
      .text(d => `${(d.value * 100).toFixed(2)}%`)
      .attr('x', d => xscale(d.name) ?? 0)
      .attr('y', d => yscale(d.value) - 2);
  }

  render() {
    return <div id="BarChart"/>;
  }
}
