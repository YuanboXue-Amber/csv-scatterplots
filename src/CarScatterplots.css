.dots circle {
  opacity: 0.5;
  fill: #f2669e;
}

.tick line {
  stroke: #C0C0BB;
}

text {
  font-family: sans-serif;
  fill: #635F5D;
}

.tick text {
  font-size: 1.5em;
}

.axisTitle text {
  font-size: 2em;
}

.chart-title text {
  font-size: 3em;
  fill: black;
}