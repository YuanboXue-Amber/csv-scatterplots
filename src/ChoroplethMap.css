#ChoroplethMap {
  padding-top: 3em;
}

.sphere {
  fill: #3bb9b9bd;
}

.country {
  /* fill: #b5f16b; */
  stroke: black;
  stroke-width: 0.15px;
}

.country:hover {
  fill: #e4747e;
}

.hilighted {
  stroke-width: 0.5px;
}