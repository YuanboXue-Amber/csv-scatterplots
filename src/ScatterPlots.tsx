import React, { Component } from 'react';
import * as d3 from 'd3';
import { isNullOrUndefined } from 'util';

export class ScatterPlots extends Component<{data: any[]}, {}> {
  componentDidMount() {
    this.draw();
  }

  // Draw scatterPlots on div with id "scatterPlots", replace if it exists
  draw() {
    const csvData = this.props.data;

    const width = 600; // svg width
    const height = 500; // svg height
    d3.select('svg').remove();
    const svg = d3
      .select('#scatterPlots') // pick div with id scatterPlots
      .append('svg')
      .attr('width', width).attr('height', height)
      .attr('align', 'center');

    const xAxisWidth = 300;
    const yAxisWidth = 300;

    // set scale for x
    let maxOfx = d3.max(csvData, (d) => d[0]);
    if (isNullOrUndefined(maxOfx)) { maxOfx = 0; }
    let minOfx = d3.min(csvData, (d) => d[0]);
    if (isNullOrUndefined(minOfx)) { minOfx = 0; }
    const xScale = d3.scaleLinear()
      .domain([1.2 * minOfx, 1.2 * maxOfx])
      .range([0, xAxisWidth]);

    // set scale for y
    let maxOfy = d3.max(csvData, (d) => d[1]);
    if (!maxOfy) { maxOfy = 0; }
    let minOfy = d3.min(csvData, (d) => d[1]);
    if (!minOfy) { minOfy = 0; }
    const yScale = d3.scaleLinear()
      .domain([1.2 * minOfy, 1.2 * maxOfy])
      .range([yAxisWidth, 0]);

    const padding = { top: 30, right: 30, bottom: 100, left: 100 };

    // draw x and y axis on svg
    const xAxis = d3.axisBottom(xScale);
    svg
      .append('g')
      .attr('class', 'axis')
      .attr(
        'transform',
        `translate(${padding.left}, ${(height - padding.bottom)})`,
      )
      .call(xAxis);
    const yAxis = d3.axisLeft(yScale);
    svg
      .append('g')
      .attr('class', 'axis')
      .attr(
        'transform',
        `translate(${padding.left}, ${(height - padding.bottom - yAxisWidth)})`,
      )
      .call(yAxis);

    // add the tooltip for mouseover dots
    const tooltip = d3.select('#scatterPlots').append('div')
      .attr('class', 'tooltip')
      .style('opacity', 0);
    // get offset of the svg element, to be used to calculate mouseover position
    let svgOffsetLeft = 0;
    let svgOffsetTop = 0;
    const svgElement = document.getElementById('scatterPlots');
    if (!isNullOrUndefined(svgElement)) {
      svgOffsetLeft = svgElement.offsetLeft;
      svgOffsetTop = svgElement.offsetTop;
    }

    // draw dots on svg
    yScale.range([0, yAxisWidth]);
    // eslint-disable-next-line
    const circle = svg
      .selectAll('circle')
      .data(csvData)
      .enter()
      .append('circle')
      .attr('fill', 'goldEnrod') // color
      .attr('cx', (d) => {
        return padding.left + xScale(d[0]);
      })
      .attr('cy', (d) => {
        return height - padding.bottom - yScale(d[1]);
      })
      .attr('r', 5) // radius
      .on('mouseover', function (d) {
          tooltip.transition()
               .style('opacity', .9);
          tooltip.text(`(${d[0]}, ${d[1]})`)
               .style('left', (parseInt(d3.select(this).attr('cx'), 10) + svgOffsetLeft) + 'px')
               .style('top', (parseInt(d3.select(this).attr('cy'), 10) + svgOffsetTop) + 'px');
      })
      .on('mouseout', (d) => {
          tooltip.transition()
               .style('opacity', 0);
      });
  }

  render() {
    this.draw();
    return <div id="scatterPlots"/>;
  }
}
