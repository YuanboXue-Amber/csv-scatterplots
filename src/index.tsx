import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
// import { BarChart } from './BarChart';
// import { RandomLetters } from './RandomLetters';
// import { CarScatterplots } from './CarScatterplots';
// import { TemperatureLineArea } from './TemperatureLineArea';
// import { FruitBowl } from './FruitBowl';
import { WorldMap } from './WorldMap';

ReactDOM.render(
  <div className="container">
    {/* <div className="container">
      <BarChart />
    </ div>
    <div className="container">
      <RandomLetters />
    </ div> */}
    {/* <div className="container">
      <CarScatterplots />
    </ div> */}
    {/* <div className="container">
      <TemperatureLineArea />
    </ div> */}
    {/* <div className="container">
      <FruitBowl />
    </ div> */}
    <div className="container">
      <WorldMap />
    </ div>
  </ div>,
  document.getElementById('root'),
);
