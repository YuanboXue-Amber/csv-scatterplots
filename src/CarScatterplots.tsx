import React, { Component } from 'react';
import * as d3 from 'd3';
import './CarScatterplots.css';
// tslint:disable-next-line: no-var-requires
const dsv = require('d3-dsv');

interface ICar {
  mpg: number;
  cylinders: number;
  displacement: number;
  horsepower: number;
  weight: number;
  acceleration: number;
  year: number;
  origin: string;
  name: string;
}

export class CarScatterplots extends Component<{}, {}> {
  componentDidMount() {
    this.draw();
  }

  async fetchFile() {
    const response = await fetch(
      'https://vizhub.com/curran/datasets/auto-mpg.csv',
    );
    // Check if the request is 200
    if (response.ok) {
      return response.text();
    }
    return Promise.reject(response);
  }

  async draw() {

    const csvString = await this.fetchFile();
    const data: d3.DSVParsedArray<ICar> = d3.csvParse(csvString, dsv.autoType);

    const title = 'Cars';
    const xvar = (d: ICar) => d.weight;
    const yvar = (d: ICar) => d.horsepower;
    const xinfo = 'weight';
    const yinfo = 'horsepower';
    const width = 960;
    const height = 500;
    const margin = {top: 80, left: 80, bottom: 80, right: 60};
    const innerwidth = width - margin.left - margin.right;
    const innerheight = height - margin.top - margin.bottom;
    const radius = 5;

    const svg = d3
      .select('#CarScatterplots')
      .selectAll('svg')
      .data([1])
      .join('svg')
        .attr('height', height)
        .attr('width', width);

    const xscale = d3
      .scaleLinear()
      .domain([d3.min(data, xvar) ?? 0, d3.max(data, xvar) ?? 0])
      .range([0, innerwidth])
      .nice();
    const yscale = d3
      .scaleLinear()
      .domain([d3.min(data, yvar) ?? 0, d3.max(data, yvar) ?? 0])
      .range([innerheight, 0])
      .nice();

    const xAxis = (g: any) => {
      const xAxisG = g.append('g');
      xAxisG
        .attr('transform', `translate(${margin.left}, ${margin.top + innerheight})`)
        .call(d3.axisBottom(xscale)
                .tickSizeInner(0 - innerheight)
                .tickPadding(10));
      xAxisG
        .select('.domain')
        .remove();
    };
    const yAxis = (g: any) => {
      const yAxisG = g.append('g');
      yAxisG
        .attr('transform', `translate(${margin.left}, ${margin.top})`)
        .call(d3.axisLeft(yscale)
                .tickSizeInner(0 - innerwidth)
                .tickPadding(5));
      yAxisG
        .select('.domain')
        .remove();
    };

    const xTitle = (g: any) => {
      const xTitleG = g.append('g');
      xTitleG
        .attr('class', 'axisTitle')
        .attr('transform', `translate(${margin.left + innerwidth / 2}, ${margin.top + innerheight + 50})`)
        .attr('text-anchor', 'middle')
        .append('text')
          .text(xinfo);
    };
    const yTitle = (g: any) => {
      const yTitleG = g.append('g');
      yTitleG
        .attr('class', 'axisTitle')
        .attr('transform', `translate(40, ${margin.top + innerheight / 2}) rotate(-90)`)
        .attr('text-anchor', 'middle')
        .append('text')
          .text(yinfo);
    };

    const chartTitle = (g: any) => {
      const titleG = g.append('g');
      titleG
        .attr('class', 'chart-title')
        .attr('transform', `translate(${margin.left + innerwidth / 2}, 50)`)
        .attr('text-anchor', 'middle')
        .append('text')
          .text(title);
    };

    svg.call(xAxis);
    svg.call(yAxis);
    svg.call(xTitle);
    svg.call(yTitle);
    svg.call(chartTitle);

    const dots = svg
      .append('g')
        .attr('class', 'dots')
        .selectAll('circle')
        .data(data)
        .join('circle')
        .attr('r', radius)
        .attr('cx', d => xscale(xvar(d)) + margin.left)
        .attr('cy', d => yscale(yvar(d)) + margin.top);
    const tooltip = svg.append('g')
      .attr('class', 'mytooltip')
      .append('text')
      .style('opacity', 0);
    dots
      .on('mouseover', function (d) {
        tooltip
          .transition()
          .style('opacity', 1);
        tooltip
          .attr('x', d3.select(this).attr('cx'))
          .attr('y', d3.select(this).attr('cy'))
          .text(`(${xvar(d)}, ${yvar(d)})`);
      })
      .on('mouseout', (d) => {
        tooltip
          .transition()
          .style('opacity', 0);
      });
  }

  render() {
    return <div id="CarScatterplots"/>;
  }
}
