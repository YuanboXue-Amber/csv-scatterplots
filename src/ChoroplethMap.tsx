import React, { Component } from 'react';
import * as d3 from 'd3';
import './ChoroplethMap.css';
import { colorLegendVertical } from './ColorLegendVertical';
// tslint:disable-next-line: no-var-requires
const topojson = require('topojson-client');

export class ChoroplethMap extends Component<{}, {}> {
  worldData: any[] = [];

  componentDidMount() {
    this.draw();
  }

  async getWorldData() {
    const [worldTopo, worldSpec]  = await Promise.all([
      d3.json('https://unpkg.com/world-atlas@1.1.4/world/50m.json'),
      d3.tsv('https://unpkg.com/world-atlas@1.1.4/world/50m.tsv'),
    ]);

    // worldGeo.id is country. worldGeo.geometry contains type Polygon and coordinates array
    const worldGeo = (topojson.feature(worldTopo, worldTopo.objects.countries)).features;

    // process both data to get worldData as:
    // worldGeo
    // worldGeo.name = worldSpec.name, worldGeo.spec = worldSpec.economy
    const map = new Map();
    worldGeo.forEach((country: any) => {
      map.set(country.id, country);
    });

    worldSpec.forEach((country: any) => {
      const countryGeo = map.get(country.iso_n3);
      countryGeo.name = country.name;
      countryGeo.spec = country.economy; // pick economy as the 'spec' we are interested in
      map.set(country.iso_n3, countryGeo);
    });
    this.worldData = Array.from(map.values());
  }

  // Draw ChoroplethMap on div with id "ChoroplethMap", replace if it exists
  async draw() {

    await this.getWorldData();

    const width = 960;
    const height = 800;
    const svg = d3
      .select('#ChoroplethMap')
      .selectAll('svg')
      .data([null])
      .join('svg')
        .attr('height', height)
        .attr('width', width);

    // map generator
    const projection = d3.geoNaturalEarth1();
    const pathGenerator = d3.geoPath(projection);

    // color scale
    const colorScale = d3.scaleOrdinal()
      .domain(this.worldData.map((d: any) => d.spec).sort());
    colorScale
      .range(d3.schemeSpectral[colorScale.domain().length]);

    // draw stable part of the map
    const mapG = svg
      .selectAll('.map')
      .data([null])
      .join('g')
        .attr('class', 'map');

    mapG
      .selectAll('.sphere')
      .data([null])
      .join('path')
        .attr('class', 'sphere')
        .attr('d', d => pathGenerator({type: 'Sphere'}));

    // zoom
    const zoom = d3.zoom()
      .scaleExtent([1, 8])
      .on('zoom', () => {
        mapG.attr('transform', d3.event.transform);
      });
    svg.call(zoom as any); // somehow, when it is mapG.call, panning by drag became really hard

    // draw color legend
    const colorLegendG = svg
      .selectAll('.colorLegend')
      .data([null])
      .join('g')
        .attr('class', 'colorLegend')
        .attr('transform', 'translate(0, 250)');
    const selector = colorLegendG;
    const colorRadius = 10;
    const colorDistance = 5;

    // add event when click color legend
    let clickedDomain: any;
    const onClick = (clicked: string) => {
      console.log(clicked);
      clickedDomain = clicked;
      renderMap();
    };

    // render dynamic part of map
    const renderMap = () => {
      const countries = mapG
        .selectAll('.country')
        .data(this.worldData)
        .join('path')
          .attr('class', 'country')
          .attr('d', (d: any) => pathGenerator(d))
          .attr('fill', (d: any) => colorScale(d.spec) as string)
          .attr('opacity', (d: any) =>
            (!clickedDomain || d.spec === clickedDomain)
            ? 1
            : 0.2,
          )
          .classed('hilighted', d => d.spec === clickedDomain);

      countries
        .selectAll('title')
        .data(d => [d])
        .join('title')
          .text((d: any) => `${d.name}: ${d.spec}`); // set hover text

      colorLegendVertical({ selector, colorScale, colorRadius, colorDistance, textwidth: 200, onClick, clickedDomain});
    };

    renderMap();
  }

  render() {
    return <div id="ChoroplethMap"/>;
  }
}
