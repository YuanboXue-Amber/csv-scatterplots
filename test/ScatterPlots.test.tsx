import React from 'react';
import { render } from '@testing-library/react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ScatterPlots } from '../src/ScatterPlots';

configure({ adapter: new Adapter() });

test('scatterPlots with input data renders correct html', () => {
  const scatterPlots = render(<ScatterPlots data={[[1.1, 1.2], [2.3, 2.5], [5.6, 7.9], [12.0, 13.4], [-3, -2.0]]} />);
  expect(scatterPlots.container).toMatchSnapshot();
});

test('scatterPlots with empty data renders correct html, has componentDidMount triggered', () => {
  const spy = jest.spyOn(ScatterPlots.prototype, 'componentDidMount');
  const scatterPlots = render(<ScatterPlots data={[]} />);
  expect(spy).toHaveBeenCalled();
  expect(scatterPlots.container).toMatchSnapshot();
});
