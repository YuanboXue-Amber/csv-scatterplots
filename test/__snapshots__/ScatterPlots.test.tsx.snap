// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scatterPlots with empty data renders correct html, has componentDidMount triggered 1`] = `
<div>
  <div
    id="scatterPlots"
  >
    <svg
      align="center"
      height="500"
      width="600"
    >
      <g
        class="axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
        transform="translate(100, 400)"
      >
        <path
          class="domain"
          d="M0.5,6V0.5H300.5V6"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(150.5,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            0
          </text>
        </g>
      </g>
      <g
        class="axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
        transform="translate(100, 100)"
      >
        <path
          class="domain"
          d="M-6,300.5H0.5V0.5H-6"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,150.5)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            0
          </text>
        </g>
      </g>
    </svg>
    <div
      class="tooltip"
      style="opacity: 0;"
    />
  </div>
</div>
`;

exports[`scatterPlots with input data renders correct html 1`] = `
<div>
  <div
    id="scatterPlots"
  >
    <svg
      align="center"
      height="500"
      width="600"
    >
      <g
        class="axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
        transform="translate(100, 400)"
      >
        <path
          class="domain"
          d="M0.5,6V0.5H300.5V6"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(27.16666666666666,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            -2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(60.49999999999999,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(93.83333333333333,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(127.16666666666667,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(160.5,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(193.83333333333331,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(227.16666666666666,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            10
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(260.5,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(293.83333333333337,0)"
        >
          <line
            stroke="currentColor"
            y2="6"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="9"
          >
            14
          </text>
        </g>
      </g>
      <g
        class="axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
        transform="translate(100, 100)"
      >
        <path
          class="domain"
          d="M-6,300.5H0.5V0.5H-6"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,294.0064935064935)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            -2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,261.538961038961)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,229.07142857142856)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,196.6038961038961)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,164.13636363636363)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,131.66883116883113)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,99.20129870129865)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            10
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,66.73376623376619)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,34.26623376623375)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            14
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,1.7987012987012547)"
        >
          <line
            stroke="currentColor"
            x2="-6"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            16
          </text>
        </g>
      </g>
      <circle
        cx="178.33333333333331"
        cy="341.5584415584416"
        fill="goldEnrod"
        r="5"
      />
      <circle
        cx="198.33333333333331"
        cy="320.45454545454544"
        fill="goldEnrod"
        r="5"
      />
      <circle
        cx="253.33333333333331"
        cy="232.79220779220776"
        fill="goldEnrod"
        r="5"
      />
      <circle
        cx="360"
        cy="143.50649350649343"
        fill="goldEnrod"
        r="5"
      />
      <circle
        cx="110"
        cy="393.5064935064935"
        fill="goldEnrod"
        r="5"
      />
    </svg>
    <div
      class="tooltip"
      style="opacity: 0;"
    />
  </div>
</div>
`;
